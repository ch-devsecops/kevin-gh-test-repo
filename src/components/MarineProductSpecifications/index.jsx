import React from 'react';
import PropTypes from 'prop-types';
import { Accordion, Box } from '@honda-canada/design-system-react';
import upperFirst from 'lodash/upperFirst';
import getValueByKey from '../../utils/getValueByKey';
import SpecificationsProvider from '../SpecificationsProvider';
import NestedAccordionTitle from '../../utils/components/NestedAccordionTitle';
import ProductSpecificationsItem from '../ProductSpecifications/ProductSpecificationsItem';

const specificationAccordionFormatter = (specifications = []) => {
  const styles = {
    containerStyles: {
      mt: 's',
      mb: 'l',
    },
    firstColumnStyles: { pl: ['s', 's', 'l'] },
    rowStyles: { maxHeight: 'unset !important' },
  };
  const categorySpecs = specifications?.map(specs => ({
    title: <NestedAccordionTitle>{upperFirst(specs?.name)}</NestedAccordionTitle>,
    key: specs?.name,
    content: <ProductSpecificationsItem styles={styles} key={specs?.trimId} specs={specs?.specs} />,
  }));
  return categorySpecs;
};

const MarineProductSpecifications = ({ fields, rendering }) => {
  if (!fields) return null;

  const gtmTags = {
    type: rendering?.componentName,
  };

  const trimId = getValueByKey(fields, 'model')[0]?.fields?.detIdentifier?.value;
  if (!trimId) return null;

  return (
    <SpecificationsProvider trims={[trimId]}>
      {({ specifications, isFetching: isFetchingSpecifications }) => {
        if (!specifications || isFetchingSpecifications) return null;
        const spec = specificationAccordionFormatter(specifications);
        return (
          <Box data-testid="cy-product-spec-wrapper">
            <Accordion
              gtmTags={gtmTags}
              items={spec}
              iconPosition="left"
              iconName="animatedArrowRightDown"
              iconColorOverride="primary"
              panelBackgroundColor="grey.3"
              borderColor="transparent"
              isCompact
              contentContainerStyles={{
                padding: 0,
                marginBottom: ['xs', 's'],
              }}
            />
          </Box>
        );
      }}
    </SpecificationsProvider>
  );
};

MarineProductSpecifications.propTypes = {
  fields: PropTypes.shape({
    detIdentifier: PropTypes.shape({
      value: PropTypes.string,
    }),
  }),
  rendering: PropTypes.shape({
    componentName: PropTypes.string,
  }),
};

export default MarineProductSpecifications;
