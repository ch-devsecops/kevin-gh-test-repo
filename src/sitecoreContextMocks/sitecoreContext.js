const commonSitecoreContext = {
  pageEditing: false,
  pageState: 'normal',
  language: 'en',
  sitecoreApiHost: 'https://app-honda-pod01-cd-plan-dmm-ccentral-dit.azurewebsites.net',
  localhostSitecoreApiHost: 'https://app-honda-pod02-cd-plan-dmm-ccentral-dit.azurewebsites.net',
  sitecoreApiKey: '453A4BB7-B80C-4351-9173-CF0748C3793F',
  hondaRestApiHost: 'uat-api.honda.ca',
  settings: { defaultProvince: 'ON' },
  sotHondaExperienceApiConfig: {
    consumerName: 'HCICONSUMERWEB',
    consumerId: '1000001',
    env: 'sit',
  },
};

const sitecoreContext = {
  engine: {
    site: {
      name: 'Engine',
    },
    ...commonSitecoreContext,
  },
  marine: {
    site: {
      name: 'Marine',
    },
    ...commonSitecoreContext,
  },
  honda: {
    site: {
      name: 'Honda',
    },
    ...commonSitecoreContext,
  },
  acura: {
    site: {
      name: 'Acura',
    },
    ...commonSitecoreContext,
  },
  mc: {
    site: {
      name: 'MC',
    },
    ...commonSitecoreContext,
  },
  psp: {
    site: {
      name: 'PSP',
    },
    partStreamConfig: {
      host: '//services.arinet.com/PartStream/',
      appKey: 'Vvsd42YoyPLGcuiWQlqj',
      id: 'aripartstream',
    },
    ...commonSitecoreContext,
  },
};

export default sitecoreContext;
