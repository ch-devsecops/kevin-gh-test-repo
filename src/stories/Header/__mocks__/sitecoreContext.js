import dictionary from '../../../sitecoreContextMocks/locales/en/translation.json';

export default {
  dictionary,
  route: {
    placeholders: {
      header: [
        {
          uid: 'aa61ffb8-7220-4df4-aa8c-0a6f9abff222',
          componentName: 'SiteLogo',
          dataSource: '{A9BA3BCE-93E8-4DA5-9DD2-552987B8691D}',
          params: {},
          fields: {
            siteLogo: {
              value: {
                src: './sample-mock-data/media/img/placeholders/logo-placeholder.png',
                alt: 'logo',
                width: '115',
                height: '37',
              },
            },
            headerUrl: {
              value: {
                href: 'http://instagram.com/hondamotoca',
                linktype: 'external',
                url: 'http://instagram.com/hondamotoca',
              },
            },
          },
        },
      ],
    },
  },
  language: 'en',
  languageSelectors: [
    {
      code: 'en',
      url: '/en/',
    },
    {
      code: 'fr',
      url: '/fr/',
    },
  ],
};
