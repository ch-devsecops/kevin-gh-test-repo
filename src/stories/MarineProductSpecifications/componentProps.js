import mockData from '../../../sample-mock-data/routes/marine-product-specifications/en.json';

export default mockData;
