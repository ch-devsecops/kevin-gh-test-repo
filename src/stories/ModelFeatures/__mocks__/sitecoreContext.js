import * as dictionary from '../../../sitecoreContextMocks/locales/en/translation.json';

const defaultSitecoreContext = {
  dictionary,
  route: {
    fields: {
      isDarkMode: {
        value: false,
      },
    },
  },
  language: 'en',
  languageSelectors: [
    {
      code: 'en',
      url: '/en/',
    },
    {
      code: 'fr',
      url: '/fr/',
    },
  ],
};

export const darkSitecoreContext = {
  ...defaultSitecoreContext,
  route: { fields: { isDarkMode: { value: true } } },
};

export default defaultSitecoreContext;
