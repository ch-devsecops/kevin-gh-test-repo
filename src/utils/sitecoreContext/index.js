export { default as useApiHost } from './useApiHost';
export { default as useApiUrl } from './useApiUrl';
export { default as useAppName } from './useAppName';
export { default as useDarkFromRoute } from './useDarkFromRoute';
export { default as useLanguage } from './useLanguage';
export { default as usePageEditing } from './usePageEditing';
export { default as usePageState } from './usePageState';
export { default as usePartStreamConfig } from './usePartStreamConfig';
export { default as useProvinces } from './useProvinces';
export { default as useProvincesPath } from './useProvincesPath';
export { default as useRoute } from './useRoute';
export { default as useSettings } from './useSettings';
export { default as useSharedApps } from './useSharedApps';
export { default as useSitecoreApiHost } from './useSitecoreApiHost';
export { default as useSitecoreApiKey } from './useSitecoreApiKey';
