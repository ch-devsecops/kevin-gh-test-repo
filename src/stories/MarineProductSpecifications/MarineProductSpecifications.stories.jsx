import React from 'react';
import MarineProductSpecifications from '../../components/MarineProductSpecifications';
import mockData from './componentProps';
import sitecoreContext from '../../sitecoreContextMocks/sitecoreContext';

export default {
  title: 'MarineProductSpecifications',
  component: MarineProductSpecifications,
};

const sitecoreContextMock = sitecoreContext.marine;

const Template = ({ storyProps }) => <MarineProductSpecifications {...storyProps} />;

export const Default = Template.bind({});
Default.args = { storyProps: mockData, context: sitecoreContextMock };
