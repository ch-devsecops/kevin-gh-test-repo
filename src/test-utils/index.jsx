import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { SitecoreContext } from '@sitecore-jss/sitecore-jss-react';
import { BrowserRouter } from 'react-router-dom';
import { OptionalVideoModalContextProvider } from '@honda-canada/design-system-react';
import UserLocationProvider from '@honda-canada/user-location';
import { configureAxe } from 'jest-axe';
import ModelExplorationContextProvider from '../components/ModelExplorationContext';
import ModelFiltersContextProvider from '../components/ModelFiltersContext';
import { MockFields, MockRendering } from './mockFields';
import MockTheme from './mockTheme';
import 'intersection-observer';
import { ACURA_THEME_NAME } from '../utils/constants';

const generateJssWrapper = context =>
  // eslint-disable-next-line react/display-name
  function ({ children }) {
    return (
      <SitecoreContext
        context={{
          language: 'en',
          site: { name: ACURA_THEME_NAME },
          ...context?.sitecoreContext,
        }}
        componentFactory={{}}
      >
        <BrowserRouter>
          <UserLocationProvider>
            <ThemeProvider theme={context?.theme || MockTheme}>
              <OptionalVideoModalContextProvider>
                <ModelExplorationContextProvider>
                  <ModelFiltersContextProvider>{children}</ModelFiltersContextProvider>
                </ModelExplorationContextProvider>
              </OptionalVideoModalContextProvider>
            </ThemeProvider>
          </UserLocationProvider>
        </BrowserRouter>
      </SitecoreContext>
    );
  };
/**
 * A custom implementation of the testing-library `render` function that
 * wraps tests in an typical JSS app's context providers.
 * Note that you can include a `context` object to set or override values
 * for SitecoreContext and ThemeContext.
 * @param {ReactElement} component - the component you want testing-library to render
 * @param {object} [options] - see https://testing-library.com/docs/react-testing-library/api/#render-options
 * @param {string} [options.context] - default values for SitecoreContext, ThemeProvider
 * @param {string} [options.context.sitecoreContext] - value for SitecoreContext
 * @param {string} [options.context.theme] - value for ThemeProvider
 * @example
 * const { getByText } = jssRender(
 *  <MyComponent />, { context: { sitecoreContext: { language: 'fr' } } }
 * );
 * @returns {function(): void} custom testing-library render function
 */
const jssRender = (component, options = {}) => {
  const { context } = options;
  // eslint-disable-next-line no-param-reassign
  delete options.context;

  return render(component, {
    wrapper: generateJssWrapper(context),
    ...options,
  });
};

const generateWrapper = context =>
  // eslint-disable-next-line react/display-name
  function ({ children }) {
    return (
      <BrowserRouter>
        <ThemeProvider theme={context?.theme || MockTheme}>
          <OptionalVideoModalContextProvider>{children}</OptionalVideoModalContextProvider>
        </ThemeProvider>
      </BrowserRouter>
    );
  };

/**
 * A custom implementation of the testing-library `render` function that
 * wraps tests in typical design-system context providers.
 * Note that you can include a `theme` object to set or override values
 * for ThemeContext.
 * @param {ReactElement} component - the component you want testing-library to render
 * @param {object} [options] - see https://testing-library.com/docs/react-testing-library/api/#render-options
 * @param {string} [options.theme] - default value for ThemeProvider
 * @example
 * const { getByText } = render(
 *  <MyComponent />, { theme: { space: { wrapperGap: '15px' } } }
 * );
 * @returns {function(): void} custom testing-library render function
 */
const customRender = (component, options = {}) => {
  const { theme } = options;
  // eslint-disable-next-line no-param-reassign
  delete options.theme;

  return render(component, {
    wrapper: generateWrapper(theme),
    ...options,
  });
};

// disable accessibility checks for errors that are usually trigged by content mistakes
const axe = configureAxe({
  rules: {
    'duplicate-id': { enabled: false },
    'heading-order': { enabled: false },
  },
});

export * from 'jest-axe';
export { axe };
export * from '@testing-library/react';
export { customRender as render, jssRender };
export { MockFields, MockRendering, MockTheme };
