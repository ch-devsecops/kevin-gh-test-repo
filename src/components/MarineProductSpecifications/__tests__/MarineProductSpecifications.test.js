import React from 'react';
import { jssRender, cleanup, axe } from 'test-utils';
import sitecoreContext from '../../../sitecoreContextMocks/sitecoreContext';
import MarineProductSpecifications from '../index';

afterEach(cleanup);

const mockData = require('../../../../sample-mock-data/routes/marine-product-specifications/en.json');
const mockEmptyData = require('../../../../sample-mock-data/routes/marine-product-specifications/empty_en.json');

const mockSpecifications = [
  {
    TrimId: parseInt(mockData.fields.data.value.model.fields.detIdentifier.value, 10),
    DisplayOrder: 1,
    label: 'SPECIFICATIONS',
    name: 'specifications',
    specs: [
      {
        displayOrder: 1,
        label: 'Engine Type',
        name: 'engine_type',
        tip: '',
        value: 'Air-cooled 4-stroke OHC',
      },
      {
        displayOrder: 2,
        label: 'Bore x Stroke',
        name: 'bore_x_stroke',
        tip: '',
        value: '56 mm x 40 mm',
      },
    ],
  },
];

jest.mock('../../SpecificationsProvider', () => ({
  __esModule: true,
  A: true,
  default: ({ children }) => (children ? children({ specifications: mockSpecifications }) : null),
}));

describe('<MarineProductSpecifications>', () => {
  test('should render correctly', () => {
    const { queryByText } = jssRender(<MarineProductSpecifications fields={mockData} />, {
      context: { sitecoreContext },
    });

    const value = queryByText('Air-cooled 4-stroke OHC');

    expect(value).toBeInTheDocument();
  });

  test('should return null if there is no fields', () => {
    const { container } = jssRender(<MarineProductSpecifications fields={null} />, { context: { sitecoreContext } });

    expect(container).toMatchSnapshot();
  });

  test('should render even if there are empty fields', () => {
    const { container } = jssRender(<MarineProductSpecifications fields={mockEmptyData} />, {
      context: { sitecoreContext },
    });

    expect(container).toMatchSnapshot();
  });

  test('should have no accessibility violations', async () => {
    const { container } = jssRender(<MarineProductSpecifications {...mockData} />, { context: { sitecoreContext } });

    expect(await axe(container)).toHaveNoViolations();
  });
});
