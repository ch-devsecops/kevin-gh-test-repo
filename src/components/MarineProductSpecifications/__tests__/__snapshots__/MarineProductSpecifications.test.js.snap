// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MarineProductSpecifications> should render even if there are empty fields 1`] = `<div />`;

exports[`<MarineProductSpecifications> should return null if there is no fields 1`] = `<div />`;
